File: bench/src/lua/fnptr.lua
local function add1(x) return x+1 end
local function mul3(x) return x*3 end
local function apply(op, x) return op(x) end
local a, b = add1, mul3
local sum = 0
for i=0,7999999 do if i%2==0 then sum = sum + apply(a, i) else sum = sum + apply(b, i) end end
os.exit(sum % 256)

File: bench/src/go/fnptr.go
package main
import "os"
type Op func(int64) int64
func add1(x int64) int64 { return x+1 }
func mul3(x int64) int64 { return x*3 }
func apply(op Op, x int64) int64 { return op(x) }
func main(){ var a Op = add1; var b Op = mul3; var sum uint64
  for i:=uint64(0);i<8000000;i++ { if i%2==0 { sum+=uint64(apply(a,int64(i))) } else { sum+=uint64(apply(b,int64(i))) } }
  os.Exit(int(sum%256)) }
